backtesting/backtesting.py

```diff
@@ -454,128 +653,248 @@ def __init__(self, *, data, cash, commission, margin, trade_on_close, length):
+
+    def new_order(self,
+                  size: float = None,
+                  limit: float = None,
+                  stop: float = None,
+                  sl: float = None,
+                  tp: float = None,
+                  *,
+                  trade: Trade = None):
+        """
+        Argument size indicates whether the order is long or short
+        """
+        size = float(size)
+        stop = stop and float(stop)
+        limit = limit and float(limit)
+        sl = sl and float(sl)
+        tp = tp and float(tp)
+
+        is_long = size > 0
+
+        if is_long:
+            # TODO: check stop as well
+            assert (sl or -np.inf) <= (limit or self.last_price) <= (tp or np.inf), \
+                "Long orders require: SL ({}) < LIMIT ({}) < TP ({})".format(
+                    sl, limit or self.last_price, tp)
+        else:
+            assert (tp or -np.inf) <= (limit or self.last_price) <= (sl or np.inf), \
+                "Short orders require: TP ({}) < LIMIT ({}) < SL ({})".format(
+                    tp, limit or self.last_price, sl)
+
+        order = Order(self, size, limit, stop, sl, tp, trade)
+        # Insert SL/TP/trade-closing orders in-front
+        if trade:
+            self.orders.insert(0, order)
+        else:
+            self.orders.append(order)
+        return order
 
     @property
-    def last_close(self):
+    def last_price(self):
         """Return price at the last (current) close.
         Used e.g. in `Orders._is_price_ok()` to see if the set price is reasonable.
         """
         return self._data.Close[-1]
 
-    def _open_position(self, price, is_long):
-        assert not self._position
-        self.orders.set_entry(None)
-
-        i, price = self._get_market_price(price)
-
-        position = float(self._cash * self._leverage / (price * (1 + self._commission)))
-        self._position = position if is_long else -position
-        self._position_open_price = price
-        self._position_open_i = i
-
-        self.log.entry_price[i] = price
-
-    def _close_position(self, price=None):
-        if not self._position:
-            return
-
-        i, price = self._get_market_price(price)
-        pl = self.position._pl(price)
-
-        self.log.pl[i] = pl
-        self.log.exit_entry[i] = self._position_open_i
-        self.log.exit_price[i] = price
-        self.log.exit_position[i] = self._position
-
-        self._cash += pl
-        self._position = 0
-        self._position_open_price = 0
-        self._position_open_i = None
-
     @property
     def equity(self):
-        return self._cash + self.position.pl
+        return self._cash + sum(trade.pl for trade in self.trades)
+
+    @property
+    def margin_available(self):
+        # From https://github.com/QuantConnect/Lean/pull/3768
+        margin_used = sum(trade.value / self._leverage for trade in self.trades)
+        return self.equity - margin_used
```

Comment:
@qacollective Thanks for the review. If you consider trying the PR further, I seem to have misplaced the offending failing/raising test at the moment, but be advised there might be a bug somewhere in/with `_Broker.margin_available()` calculation. :sweat_smile:

Reply:
And not necessarily the same, but related: https://github.com/kernc/backtesting.py/issues/77.

Reply:
@kernc No worries!!  I'm a bit new to GitHub as well, so should have submitted a proper review with suggested code change which I see is now possible from here.  

FYI I really liked the look / feel of backtesting.py when I first saw it, but the fact that it allocated 100% of the funds and didn't seem to be able to hold multiple concurrent orders turning into multiple simultaneous positions rendered it a non-candidate for my use until I stumbled across this PR - which on my current understanding, changes all of those things in one big hit :ok_hand:

At the moment the only thing that I'm not new to here is Python.  So I'm quite tentative.  I'm in the early days of getting to understand backtesting.py so I can run it over some 100% fake but mathematically pure and simple data and strategy to test that it works to my understanding and requirements.  I will let you know if I find anything that doesn't look right! :thinking:

Reply:
> so I can run it over some 100% fake but mathematically pure and simple data and strategy to test that it works to my understanding and requirements. I will let you know if I find anything that doesn't look right! thinking

Even much less than that would be seriously awesome and very appreciated. :smiling_face_with_three_hearts: